allennlp/training/optimizers.py

```diff
@@ -196,6 +205,159 @@ def default(model_parameters: List) -> "Optimizer":
+
+        self._grouped_optimizers: Dict[str, Optimizer] = {}
+
+        optimizer_names = [optimizer["name"] for optimizer in self.optimizers]
+        assert "default" in optimizer_names, (
+           "No optimizer was provided for the 'default' group."
+           " Please provide a 'name': 'default' parameter for one of the `optimizers`."
+           )
+
+        parameter_groups = make_parameter_groups(self.model_parameters, self.parameter_groups)
+        # Manually set optimizer kwargs for the default group, so certain schedulers are able
+        # to know the options for this group.
+        if len(list(parameter_groups[-1].keys())) == 1:
+            for optimizer in self.optimizers:
+                if optimizer["name"] == "default":
+                    parameter_groups[-1].update(optimizer)
+
+        optimizer_groups = self._populate_optimizer_groups(parameter_groups)
+        self._set_optimizer_kwargs(optimizer_groups, self.optimizers)
+
+        # Check to see if an optimizer didn't receive any parameters.
+        for optimizer_key, optimizer_parameters in optimizer_groups.items():
+            if len(optimizer_parameters[0]) == 0:
+               raise ValueError(
+                    f"Optimizer '{optimizer_key}' did not receive any parameters."
+                    " If you are using `parameter_groups`, please make sure that the regexes you have provided"
+                    " match the desired model parameters, or that the `name` value of this optimizer "
+                    " matches that of the parameter group you are trying to assign to it."
+                    " Alternatively, you can remove this optimizer from the provided `optimizers`"
+                    " if it is not relevant to a particular parameter group."
+                    )
+
+        # We have already created our parameter groups so we set `parameter_groups` to False, which is a no-op in `make_parameter_groups`.
+        # This is so that we don't have to change any of the code when initializing individual optimizers.
+        for optimizer_name, (params, optimizer_kwargs) in optimizer_groups.items():
+            self._grouped_optimizers[optimizer_name] = Optimizer.from_params(model_parameters=params, parameter_groups=False, params=Params(optimizer_kwargs))
+        
+        super().__init__(parameter_groups, optimizer_groups["default"][1])
+
+    def _populate_optimizer_groups(self, parameter_groups):
+        """
+        Assigns each parameter group in `parameter_groups` to the
+        appropriate optimizer or to the 'default' optimizer.
+        """
+        optimizer_groups = {}
+
+        for parameter_group in parameter_groups:
+            # Check to see what optimizer this group should be assigned to.
+            if "name" in list(parameter_group.keys()):
+                optimizer_key = parameter_group["name"]
+                if not optimizer_key in optimizer_groups:
+                    optimizer_groups[optimizer_key] = ([], {})
```

Comment:
Each group contains a tuple where the first element is an iterable which stores model parameters or groups of parameters and the second element is a dictionary for the optimizer's default options. For reference: https://pytorch.org/docs/stable/optim.html#constructing-it